### PR TITLE
(SR-8830) implement generic packing/unpacking in irgen using `TensorGroup` protocol

### PR DESCRIPTION
**Pull request by marcrasi**
Adds a `TensorGroup` protocol, and teaches IRGen to use it for packing and unpacking.

This is the beginning of the generic packing/unpacking plan described in https://bugs.swift.org/browse/SR-8830.

Plus a few other tangential things:
* Makes TensorHandle.cTensorHandle and TensorHandle.init(owning cTensorHandle) public so that conformances to `TensorGroup` can be implemented. Prepends some underscores to the names.
* Yet another fix to the `SILFunction.processedByDeabstraction` flag. The problem I encountered is that the optimizer specializes deabstracted functions, and does not know to set the flag on the resulting specialized functions. This makes IRGen incorrectly bail out when IRGen'ing the specialized functions.
* Replaced our `createStringValAddr` function with an existing `getAddrOfGlobalString` that does the same thing except with deduplication.

### CODE REVIEW

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 1]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+func _TFCOpAddInputFromTensorHandle(_ op: CTFEOp,
+                                    _ handle: _AnyTensorHandle,
+                                    _ status: CTFStatus) {
+  TFE_OpAddInput(op, handle.cTensorHandle, status)
+}
+
+/// Adds `t` as an input or inputs to `op`.
+@usableFromInline
+@_silgen_name("_swift_tfc_OpAddInputFromTensorGroup")
+func _TFCOpAddInputFromTensorGroup<T: TensorGroup>(
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Nit: In the stdlib codebase we tend to use `T : P` instead of `T: P`. Same for other places.

```suggestion
func _TFCOpAddInputFromTensorGroup<T : TensorGroup>(
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 2]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+  }
+  return count
+}
+
+/// Initializes a TensorGroup value, taking ownership of all the tensor
+/// handles in `tensorHandles`. Also deallocates `tensorHandles`.
+@usableFromInline
+@_silgen_name("_swift_tfc_InitTensorGroup")
+func _TFCInitTensorGroup<T: TensorGroup>(
+    owning tensorHandles: UnsafeMutablePointer<CTensorHandle>
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: I'd suggest dropping argument labels for top-level, runtime entry functions, because they are no direct callers from Swift.

File: stdlib/public/TensorFlow/TensorGroup.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
+//===-- TensorGroup.swift ---------------------------------*- swift -*-===//
```
Comment (rxwei) [inline stdlib/public/TensorFlow/TensorGroup.swift]: Please extend this line to fill all 80 columns.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 4]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+func _TFCOpAddInputFromTensorGroup<T: TensorGroup>(
+    _ op: CTFEOp, _ t: T, _ status: CTFStatus
+) -> Int32 {
+  let count = T._tensorHandleCount
+  let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: Int(count))
+  defer { buffer.deallocate() }
+  t._unpackTensorHandles(resultBuffer: buffer)
+  for i in 0..<count {
+    TFE_OpAddInput(op, buffer.advanced(by: Int(i)).pointee, status)
+    if TF_GetCode(status) != TF_OK {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Use `guard TF_GetCode(status) == TF_OK else` instead of `if`.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 5]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+  TFE_OpAddInput(op, handle.cTensorHandle, status)
+}
+
+/// Adds `t` as an input or inputs to `op`.
+@usableFromInline
+@_silgen_name("_swift_tfc_OpAddInputFromTensorGroup")
+func _TFCOpAddInputFromTensorGroup<T: TensorGroup>(
+    _ op: CTFEOp, _ t: T, _ status: CTFStatus
+) -> Int32 {
+  let count = T._tensorHandleCount
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Since the `Int32` version of `count` is never used in this function (as far as i can tell), you might as well convert it to `Int` here and drop all other conversions below.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 6]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+  let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: Int(count))
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Instead of using an `UnsafeMutablePointer` and manually maintaining a `count`, I would suggest using `UnsafeMutableBufferPointer.allocate(capacity:)`, and you can then do a `for`-loop over it without having touching `count` again.

File: stdlib/public/TensorFlow/TensorHandle.swift [hunk 7]
```diff
@@ -26,7 +26,7 @@ public class _AnyTensorHandle {
   /// - Note: The compiler knows that `_AnyTensorHandle` has a single stored
   /// property, and assumes that this is it. Changing the design of
   /// `TensorHandle` will require tweaking the compiler.
-  @usableFromInline let cTensorHandle: CTensorHandle
+  public let cTensorHandle: CTensorHandle
```
Comment (rxwei) [inline stdlib/public/TensorFlow/TensorHandle.swift]: There's no good solutions to this other than prepending an underscore to it.

File: stdlib/public/TensorFlow/TensorHandle.swift [hunk 8]
```diff
@@ -42,8 +42,7 @@ public class _AnyTensorHandle {
 @_fixed_layout // required because the compiler accesses cTensorHandle directly.
 public final class TensorHandle<Scalar> : _AnyTensorHandle
   where Scalar : AccelerableByTensorFlow {
-  @usableFromInline
-  init(owning cTensorHandle: CTensorHandle) {
+  public init(owning cTensorHandle: CTensorHandle) {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/TensorHandle.swift]: I'd suggest prepending an underscore to the first argument label because this should never be accessed by the user or show up in code completion. Also feel free to file a bug to do this after this patch lands.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 9]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+  let t = T(_owning: tensorHandles)
+  tensorHandles.deallocate()
+  return t
+}
+
+/// Allocates a buffer with enough capacity to hold tensor operation results
+/// corresponding to a TensorGroup of type `T`.
+@usableFromInline
+@_silgen_name("_swift_tfc_GetTensorGroupResultBuffer")
+func _TFCGetTensorGroupResultBuffer<T: TensorGroup>(
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Some comments on naming:
First, this name does not reflect the fact that it's doing a heap allocation.
Second, it's unclear what "result buffer" means except when it's associated with the `_unpack` method.

Maybe call it `_TFCAllocateTensorGroupCHandleBuffer`?

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 10]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+  _ type: T.Type
+) -> UnsafeMutablePointer<CTensorHandle> {
+  return UnsafeMutablePointer.allocate(capacity: Int(T._tensorHandleCount))
+}
+
+/// Returns the number of tensor operation results required to initialize a
+/// TensorGroup of type `T`.
+@_silgen_name("_swift_tfc_GetTensorGroupResultBufferSize")
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: `_TFCGetTensorGroupCHandleCount`?

File: lib/IRGen/IRGenSIL.cpp [hunk 11]
```diff
@@ -2024,43 +2012,79 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
+      checkOk(status);
+      return llvm::ConstantInt::get(IGM.Int32Ty, 1);
+    }
+
+    auto canType = opInput->getType().getASTType()->getCanonicalType();
+    auto conformance = tfModule->lookupConformance(canType,
+                                                   tensorGroupProto);
+    assert(conformance && "input type does not conform to TensorGroup");
+    auto *typeMetadata = emitTypeMetadataRef(canType);
+    auto *wtable = emitWitnessTableRef(*this, canType, *conformance);
```
Comment (rxwei) [inline lib/IRGen/IRGenSIL.cpp]: Nice!

File: stdlib/public/TensorFlow/TensorGroup.swift [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+/// outputs. When a TensorGroup is used as an input, it gets passed to the
+/// tensor operation as an input list whose elements are the tensor fields of
+/// the type. When a TensorGroup is used as an output, it gets initialized
+/// with its tensor fields set to the tensor operation's tensor outputs.
+///
+/// TODO: Add a derived conformance to TensorGroup so that users don't have
+/// to write the conformance themselves.
+public protocol TensorGroup {
+  /// The number of tensor fields in this type.
+  static var _tensorHandleCount: Int32 { get }
```
Comment (mhong) [inline stdlib/public/TensorFlow/TensorGroup.swift]: why do we need this property, instead of having caller just use `_unknownShapeList.count`?

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 13]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+/// Adds `t` as an input or inputs to `op`.
+@usableFromInline
+@_silgen_name("_swift_tfc_OpAddInputFromTensorGroup")
+func _TFCOpAddInputFromTensorGroup<T: TensorGroup>(
+    _ op: CTFEOp, _ t: T, _ status: CTFStatus
+) -> Int32 {
+  let count = T._tensorHandleCount
+  let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: Int(count))
+  defer { buffer.deallocate() }
+  t._unpackTensorHandles(resultBuffer: buffer)
```
Comment (mhong) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: question on API design: would it be better to instead have _unpackTensorHandles() allocate and return the buffer (e.g. `UnsafeMutablePointer<CTensorHandle>`)? this way we don't need to worry about `resultBuffer` being too small.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 14]
```diff
@@ -1172,6 +1172,65 @@ public func _TFCGetCTensorHandleFromSwift(
...
+/// Adds `handle` as an input to `op`.
+@usableFromInline
+@_silgen_name("_swift_tfc_OpAddInputFromTensorHandle")
+func _TFCOpAddInputFromTensorHandle(_ op: CTFEOp,
+                                    _ handle: _AnyTensorHandle,
+                                    _ status: CTFStatus) {
+  TFE_OpAddInput(op, handle.cTensorHandle, status)
+}
+
+/// Adds `t` as an input or inputs to `op`.
```
Comment (mhong) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: document return value

File: lib/IRGen/IRGenSIL.cpp [hunk 15]
```diff
@@ -2024,43 +2012,79 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
+  // are a few different cases that can be unpacked:
+  // - if `opInput` is a TensorFlow value, then we just add its handle;
+  // - if `opInput` is an archetype conforming to TensorGroup, then we
+  //   ask the conformance for the handles and add those;
+  // This function crashes if it receives an unhandled case. Earlier
+  // typechecking should ensure that inputs match the cases that this function
+  // handles.
+  //
+  // TODO: We should also handle the following cases:
+  // - if `opInput` is an array of TensorFlow values, then we add all the
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: so we handled the struct case above. what about tuples?

File: lib/IRGen/IRGenSIL.cpp (hunk 11 above)
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: +1

File: lib/IRGen/IRGenSIL.cpp [hunk 16]
```diff
@@ -2024,43 +2012,79 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
-    auto cHandle =
-        Builder.CreateCall(extractHandleFn, {tensorHandleValueUntyped});
 
-    // Add an op input as in:
-    //   TFE_OpAddInput(op, cHandle);
-    auto *opAddInputFn = IGM.getTFE_OpAddInputFn();
-    Builder.CreateCall(opAddInputFn, {op, cHandle, status});
+    auto *opAddInputFromTensorGroupFn =
+        IGM.getTFC_OpAddInputFromTensorGroupFn();
+    auto *ret = Builder.CreateCall(
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: nit: ret -> numInputs?

File: lib/IRGen/IRGenSIL.cpp [hunk 17]
```diff
@@ -2024,43 +2012,79 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
   };
 
+  // If we find an out parameter, we'll keep track of its address and some data
+  // about its type.
+  Address outParameterAddress;
+  llvm::Value *outParameterTypeMetadata = nullptr;
+  llvm::Value *outParameterTensorGroupWitnessTable = nullptr;
+
   // Handle inputs and dynamic attributes.
+  unsigned inputIdx = 0;
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: with the packing and unpacking going on, the semantics `inputIdx` could become unclear, and the associated code could be harder to read and maintain.

consider adding an example, and maybe introduce some additional terminology. one option can be (pls feel free to come up with sth else):

say the op takes 3 _input params_ (with additional attrs but we don't consider them input), where each param can be a single tensor handle or a list. Say this op takes: (handle, list, handle), and the list in this case contains 3 _elements_.

We first unpack these params into a flat list of 5 tensor handle _elements_. For the middle _param_, it has an attr N, and we set it to be 3.

File: lib/IRGen/IRGenSIL.cpp [hunk 18]
```diff
@@ -2074,31 +2098,69 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
+      // Special attr to represent the list length, if needed. This is necessary
+      // because we add list inputs one element at a time rather than by using a
+      // TFE_OpAddInputList function (which doesn't exist yet). This is null
+      // when the input does not have a number attr.
+      const char *numberAttr = inputListNumberAttr(opInfo.getOperationName(),
+                                                   inputIdx);
+
+      // Keep track of the list length when we have a number attr.
+      llvm::Value *listLength = nullptr;
+      auto incrementListLength = [&](llvm::Value *amount) {
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: the code here feels pretty tricky...

one option to simplify it is: make incrementListLength only callable from a list context, and make the caller initialize the list length first before calling this closure, so that here we only do `Builder.CreateAdd(listLength, amount);`

also, i wonder if the logic is correct if we have mixed input params with handles and lists. say the tfop takes 2 params: (list, handle). IIUC, say we have the N attr, and user writes: `#tfop("SomeOp", x, [y, z])`, where x, y and z are tensor handles (this example does not involve the new TensorGroup). In this case, will `listLength` end up being 3, instead of 2?

File: lib/IRGen/IRGenSIL.cpp [hunk 19]
```diff
@@ -2074,31 +2098,69 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
       assert(0 && "TensorAttributes cannot be dynamic");
     case GraphOperationInfo::ArgumentLowering::Input: {
       assert(argumentName.empty() && "inputs cannot have names");
+
+      // Special attr to represent the list length, if needed. This is necessary
+      // because we add list inputs one element at a time rather than by using a
+      // TFE_OpAddInputList function (which doesn't exist yet). This is null
+      // when the input does not have a number attr.
+      const char *numberAttr = inputListNumberAttr(opInfo.getOperationName(),
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: is there an op that can take 2 or more input lists? say the op takes total of 3 input params, and they are (list, handle, list). 

if yes:
1. in that case, would there be two "numberAttr" attrs, maybe named "N1" and "N2"?
2. i guess the code here wouldn't generalize to those cases (if so, pls feel free to address later -- this patch is already very large).

**Review by mhong — COMMENTED** [review]
Nice patch!

Many comments are on style, but a few are on design/correctness. Please feel free to address the latter ones first.

**Comment by mhong** [comment]
Do you want to link to https://bugs.swift.org/browse/SR-8830 in the PR description?

File: stdlib/public/TensorFlow/TensorGroup.swift [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+  /// An array of `nil`s with size equal to `_tensorHandleCount`.
+  static var _unknownShapeList: [TensorShape?] { get }
+
+  /// The types of the tensor fields in this type.
+  static var _typeList: [TensorDataType] { get }
+
+  /// Writes the tensor handles to `resultBuffer`, which must be allocated
+  /// with enough capacity to hold `_tensorHandleCount` handles. The returned
+  /// tensor handles are borrowed: this container still owns them.
+  func _unpackTensorHandles(resultBuffer: UnsafeMutablePointer<CTensorHandle>)
```
[the remark this answers is not in the document]
Comment (rxwei) [reply stdlib/public/TensorFlow/TensorGroup.swift]: +1 to "into".

File: include/swift/Runtime/RuntimeFunctions.def [hunk 21]
```diff
@@ -1418,6 +1418,36 @@ FUNCTION(TFC_GetCTensorHandleFromSwift, _swift_tfc_GetCTensorHandleFromSwift, C_
          ARGS(Int8PtrTy),
          ATTRS(NoUnwind))
 
+FUNCTION(TFC_OpAddInputFromTensorHandle, _swift_tfc_OpAddInputFromTensorHandle,
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/Runtime/RuntimeFunctions.def]: I'll delete the RuntimeFunctions.def entry. There is a test that calls TFC_GetCTensorHandleFromSwift: https://github.com/apple/swift/blob/a51ad275afeed7716306829289830f30f19e6279/test/TensorFlowRuntime/runtime_entry_points.swift. Though it should be easy to modify the test to not call that. I'll modify the test and delete TFC_GetCTensorHandleFromSwift.

File: lib/IRGen/IRGenSIL.cpp [hunk 22]
```diff
@@ -1941,12 +1922,13 @@ llvm::Value *IRGenSILFunction::createArrayAndSize(
 /// FIXME: This is incomplete. We should derive this from op metadata. We can
 /// get it from the "number_attr" field of the input's "input_arg". The raw op
 /// generator is probably the right place to put this.
-static const char *inputListNumberAttr(StringRef opName) {
-  if (opName == "Pack")
+static const char *inputListNumberAttr(StringRef opName, unsigned inputIdx) {
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: I added a comment explaining in more detail what this attr is for and what inputIdx does.

In the process, I discovered that I was handling "Concat" wrong. It's input list has `inputIdx == 1`. So I fixed that.

Since this is like a database of ops, I think it makes most sense to structure it as a big if-elseif statement that figures out what op it is, and then check the `inputIdx` inside each body to decide what to return.

File: lib/IRGen/IRGenSIL.cpp (hunk 17 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: I added a comment with an example that should clarify what is happening.

File: lib/IRGen/IRGenSIL.cpp (hunk 19 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: Yes, there are such ops. For example, the "ParseExample" example that I added in the comments. The code should already handle such cases (as long as we add the ops to the hardcoded database in `getInputListNumberAttr`).

File: lib/IRGen/IRGenSIL.cpp (hunk 18 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: I simplified it to construct the value immediately, so that the closure only has to CreateAdd.

I cannot restrict it to only the SAK_List case, because once I add support for runtime arrays, the runtime arrays will show up as SAK_Singles.

The logic should be correct for mixed handles and lists. For example `#tfop("SomeOp", x, [y, z])` gets 2 `structuredArgument`s: one for `x` and one for `[y, z]`. Then it calculates separate `listLength`s for them in separate loop iterations. So it should get `listLength == 1` for `x` and `listLength == 2` for `[y, z]`.

File: stdlib/public/TensorFlow/TensorHandle.swift (hunk 7 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/TensorHandle.swift]: I prepended an underscore.

File: stdlib/public/TensorFlow/TensorHandle.swift (hunk 8 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/TensorHandle.swift]: I prepended an underscore.

**Review by marcrasi — COMMENTED** [review]
I addressed all the comments.

File: lib/IRGen/IRGenSIL.cpp [hunk 23]
```diff
@@ -2693,28 +2821,45 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
+    auto *returnValuesAddressUntyped = Builder.CreateBitCast(
+        returnValuesAddress, IGM.Int8PtrTy);
+    Builder.CreateCall(initTensorGroupFn,
+                       {outParameterAddressOpaque,
+                        returnValuesAddressUntyped,
+                        outParameterTypeMetadata,
+                        outParameterTensorGroupWitnessTable});
+  } else {
+    // There is no out parameter. Return the results directly.
+    for (unsigned resultIdx = 0; resultIdx < i->getNumResults(); ++resultIdx) {
```
Comment (rxwei) [inline lib/IRGen/IRGenSIL.cpp]:
```suggestion
    for (unsigned resultIdx : range(i->getNumResults())) {
```

File: lib/IRGen/IRGenSIL.cpp [hunk 24]
```diff
@@ -2693,28 +2821,45 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
+    auto *outParameterAddressOpaque = Builder.CreateBitCast(
+        outParameterAddress.getAddress(), IGM.OpaquePtrTy);
+    auto *returnValuesAddressUntyped = Builder.CreateBitCast(
+        returnValuesAddress, IGM.Int8PtrTy);
+    Builder.CreateCall(initTensorGroupFn,
+                       {outParameterAddressOpaque,
+                        returnValuesAddressUntyped,
+                        outParameterTypeMetadata,
+                        outParameterTensorGroupWitnessTable});
+  } else {
```
Comment (rxwei) [inline lib/IRGen/IRGenSIL.cpp]: Since this sits directly above a return, would it be nicer to do an early return in the if-branch and out-scope the else-branch?

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 25]
```diff
@@ -1164,12 +1164,64 @@ func _TFCGetGlobalEagerContext() -> CTFEContext {
...
+@usableFromInline
+@_silgen_name("_swift_tfc_OpAddInputFromTensorGroup")
+func _TFCOpAddInputFromTensorGroup<T : TensorGroup>(
+    _ op: CTFEOp, _ t: T, _ status: CTFStatus
+) -> Int32 {
+  let count = Int(T._tensorHandleCount)
+  let buffer =
+      UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: count)
+  defer { buffer.deallocate() }
+  t._unpackTensorHandles(into: buffer.baseAddress!)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Please add a test for an empty struct, in which case `baseAddress` may be `nil`.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 26]
```diff
@@ -1164,12 +1164,64 @@ func _TFCGetGlobalEagerContext() -> CTFEContext {
...
+/// Adds `handle` as an input to `op`.
+@usableFromInline
+@_silgen_name("_swift_tfc_OpAddInputFromTensorHandle")
+func _TFCOpAddInputFromTensorHandle(_ op: CTFEOp,
+                                    _ handle: _AnyTensorHandle,
+                                    _ status: CTFStatus) {
+  TFE_OpAddInput(op, handle._cTensorHandle, status)
+}
+
+/// Adds `t` as an input or inputs to `op`. Returns the number of inputs added.
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Please document what the return value means.

**Review by rxwei — APPROVED** [review]
Good work!

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 26 above)
Comment (rxwei) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: Ah, my bad, i didn't notice that the comment is right there :)

File: include/swift/Runtime/RuntimeFunctions.def (hunk 21 above)
[the remark this answers is not in the document]
Comment (mhong) [reply include/swift/Runtime/RuntimeFunctions.def]: Please feel free to leave a TODO for deleting old entries, if doing that in another patch is easier.

File: include/swift/SIL/SILFunction.h [hunk 27]
```diff
@@ -346,13 +346,14 @@ class SILFunction
...
-  /// Whether TFDeabstraction processed this function.
+  /// Whether TFDeabstraction specifically decided to not to deabstract this
+  /// function.
   ///
   /// This is a temporary flag that IRGen uses to determine if the graph_ops in
   /// this function are safe to lower. Once IRGen is powerful enough to lower
   /// every graph_op that it receives, we should remove this flag.
   /// TODO: Remove this flag.
-  bool processedByDeabstraction = false;
+  bool skippedByDeabstraction = false;
```
[the remark this answers is not in the document]
Comment (mhong) [reply include/swift/SIL/SILFunction.h]: SGTM

File: lib/IRGen/IRGenModule.cpp [hunk 28]
```diff
@@ -561,7 +569,7 @@ llvm::Constant *swift::getRuntimeFn(llvm::Module &Module,
     for (auto Attr : attrs) {
       if (isReturnAttribute(Attr))
         buildRetAttr.addAttribute(Attr);
-      else if (isReturnedAttribute(Attr))
+      else if (isReturnedAttribute(Attr) || isStructRetAttribute(Attr))
```
Comment (mhong) [inline lib/IRGen/IRGenModule.cpp]: add `// SWIFT_ENABLE_TENSORFLOW` here as well?

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 29]
```diff
@@ -449,6 +451,15 @@ static GraphOperationInst *simplifyOperands(GraphOperationInst *origInst,
       auto argumentType = argumentValue->getType();
       assert(argumentType.isAddress());
       if (!argumentType.isLoadable(origInst->getModule())) {
+        // When we're in dynamic compilation mode, it is okay to keep the
+        // generic out parameter because dynamic compilation knows how to deal
+        // with generic out parameters.
+        if (llvm::TFDynamicCompilation && !isAcceleratorOnly(*fn)) {
+          opBuilder.addArgument(argumentValue,
```
[the remark this answers is not in the document]
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Great!

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 30]
```diff
@@ -2316,16 +2345,26 @@ void TFDeabstraction::evaluateAttributesAndDoPacking(
         // Add each element to the input list we're building, drilling down
         // through any aggregates that may be around it.
         for (auto *use : arrayDecoder.elementsAtInit) {
-          auto *store = dyn_cast<StoreInst>(use->getUser());
-          if (!store) {
+          SILValue element;
+          if (auto *store = dyn_cast<StoreInst>(use->getUser()))
+            element = store->getSrc();
+          if (auto *copyAddr = dyn_cast<CopyAddrInst>(use->getUser()))
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: nit: use `else if` here.

File: stdlib/public/TensorFlow/TensorGroup.swift (hunk 12 above)
Comment (mhong) [reply stdlib/public/TensorFlow/TensorGroup.swift]: can we remove this stored property, and instead use `_typeList.count`? 

(we can make this a computed property if that makes caller's code simpler somehow.)

File: lib/IRGen/IRGenSIL.cpp [hunk 31]
```diff
@@ -1936,18 +1917,46 @@ llvm::Value *IRGenSILFunction::createArrayAndSize(
   return numElts;
 }
 
-/// The name of the number attr that describes the length of the input list to
-/// the op. Returns nullptr if no number attr is required.
+/// The name of the numberAttr for `opName`'s `inputIdx` argument. The
+/// numberAttr is an attr whose value must be set to describe the length of an
+/// input list. Returns nullptr if no number attr is required. For example,
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: nit: "number attr" -> "numberAttr"

File: lib/IRGen/IRGenSIL.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: Looks great!

File: lib/IRGen/IRGenSIL.cpp (hunk 15 above)
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: Makes sense. Is there ever a case where a tuple is not statically known?

File: lib/IRGen/IRGenSIL.cpp (hunk 17 above)
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: very nice

File: lib/IRGen/IRGenSIL.cpp (hunk 18 above)
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: SG overall.

> I cannot restrict it to only the SAK_List case, because once I add support for runtime arrays, the runtime arrays will show up as SAK_Singles.

hmm i feel this case might cause confusion. We'll need to take care (e.g. add more comments on the GraphOperationInfo::SAK_Single enum case if needed)

File: lib/IRGen/IRGenSIL.cpp [hunk 32]
```diff
@@ -2074,31 +2153,62 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
...
+      auto incrementListLength = [&](llvm::Value *amount) {
+        listLength = Builder.CreateAdd(listLength, amount);
+      };
+
       switch (structuredArgument.getKind()) {
       case GraphOperationInfo::SAK_Single: {
         LLVM_DEBUG(llvm::dbgs() << " Adding a single input.\n");
         auto tensorHandle = structuredArgument.getSingleArgument();
-        addTensorhandleAsOpInput(tensorHandle);
+        incrementListLength(unpackAndAddInput(tensorHandle));
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: i still find it a bit unintuitive and complex that we are calling `incrementListLength()` in this context, but there is no loop around it. It'd probably be cleaner if we instead call another function here, sth like `setListLength()`. this way we can probably also move `llvm::Value *listLength = llvm::ConstantInt::get(IGM.Int32Ty, 0);` to the `case GraphOperationInfo::SAK_List` block below. 

But my preference is not very strong here.

**Review by mhong — APPROVED** [review]
Great work! LGTM.

Just a few more comments on style and suggestions on code simplification.

File: lib/IRGen/IRGenSIL.cpp (hunk 24 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: I do not think early return makes sense here. It's two "equally successful" cases, not one case where we want to bail out and another case where we want to keep doing a bunch more processing.

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 25 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: good catch. I added a test and it seems to pass, but I'll change `_unpackTensorHandles` to take an `UnsafeMutablePointer?` anyways, so that we only do force-unwrapping when we're sure that the struct is nonempty.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 30 above)
Comment (marcrasi) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: done

File: stdlib/public/TensorFlow/TensorGroup.swift (hunk 12 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/TensorGroup.swift]: done

File: lib/IRGen/IRGenModule.cpp (hunk 28 above)
Comment (marcrasi) [reply lib/IRGen/IRGenModule.cpp]: done

File: lib/IRGen/IRGenSIL.cpp (hunk 31 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: done

File: lib/IRGen/IRGenSIL.cpp (hunk 32 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: I changed it to set it directly in the non-loop case

File: lib/IRGen/IRGenSIL.cpp (hunk 23 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: done

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/IRGen/IRGenSIL.cpp (hunk 32 above)
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: love the new code structure!
